atoma-state/src/handlers.rs

```diff
@@ -1827,4 +1902,176 @@ pub mod remote_attestation_verification {
+            tracing::info!(
+                target = "atoma-state-handlers",
+                event = "attest-nvidia-evidence-list",
+                "Attestation successful for device type: {device_type} and public key: {}",
+                hex::encode(new_public_key),
+            );
+        } else {
+            tracing::error!(
+                target = "atoma-state-handlers",
+                event = "attest-nvidia-evidence-list",
+                "Attestation failed for device type: {device_type} and public key: {}",
+                hex::encode(new_public_key),
+            );
+        }
+
+        Ok(())
+    }
+
+    /// Verifies the topology of the GPU and NVSwitch evidence data
+    ///
+    /// This function verifies the topology of the GPU and NVSwitch evidence data by decoding the evidence data and checking the topology.
+    ///
+    /// # Arguments
+    ///
+    /// * `gpu_evidence_data` - A reference to the GPU evidence data
+    /// * `nvswitch_evidence_data` - A reference to the NVSwitch evidence data
+    ///
+    /// # Returns
+    ///
+    /// * `Result<()>` - Ok(()) if the topology is valid, or an error if the operation failed
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if:
+    /// * The GPU evidence data cannot be decoded
+    /// * The NVSwitch evidence data cannot be decoded
+    /// * The topology check fails
+    #[instrument(level = "info", skip_all, fields(topology_check = "true",), err)]
+    pub fn verify_topology(
+        gpu_evidence_data: &[DeviceEvidence],
+        nvswitch_evidence_data: &[NvSwitchEvidence],
+    ) -> Result<()> {
+        let gpu_evidence_data_refs = gpu_evidence_data
+            .iter()
+            .map(|evidence| STANDARD.decode(&evidence.evidence))
+            .collect::<std::result::Result<Vec<_>, _>>()
+            .map_err(|e| {
+                tracing::error!(
+                    target = "atoma-state-handlers",
+                    event = "verify-topology",
+                    "Failed to decode GPU evidence data: {e}",
+                );
+                AtomaStateRemoteAttestationError::FailedToAttestRemote(
+                    AttestError::RemoteAttestationFailed,
+                )
+            })?;
+        let nvswitch_evidence_data_refs = nvswitch_evidence_data
+            .iter()
+            .map(|evidence| STANDARD.decode(&evidence.evidence))
+            .collect::<std::result::Result<Vec<_>, _>>()
+            .map_err(|e| {
+                tracing::error!(
+                    target = "atoma-state-handlers",
+                    event = "verify-topology",
+                    "Failed to decode NVSwitch evidence data from base64 data: {e}",
+                );
+                AtomaStateRemoteAttestationError::FailedToAttestRemote(
+                    AttestError::RemoteAttestationFailed,
+                )
+            })?;
+        let unique_switch_pdis_set = topology::topology::gpu_topology_check(
```

Comment:
piggy backing off the suggestion above we could also preserve the error returned in the `AtomaStateRemoteAttestationError::FailedToAttestRemote`

```rust
// Check GPU topology
    let unique_switch_pdis_set = topology::topology::gpu_topology_check(&gpu_evidence_slices)
        .map_err(|e| {
            error!("GPU topology check failed: {}", e);
            AtomaStateRemoteAttestationError::TopologyCheckFailed { 
                check_type: "GPU".to_string(),
                source: Box::new(e)
            }
        })?;
    
    ```